File: src/main/java/xyz/phanta/cbtweaker/structure/StructureMatcher.java
package xyz.phanta.cbtweaker.structure;

import net.minecraft.util.math.BlockPos;
import net.minecraft.util.math.Vec3i;
import net.minecraft.world.World;
import org.apache.commons.lang3.tuple.Pair;
import xyz.phanta.cbtweaker.structure.block.StructureBlockMatcher;
import xyz.phanta.cbtweaker.util.Direction;

import javax.annotation.Nullable;
import java.util.Collection;
import java.util.Iterator;

public interface StructureMatcher {

    Iterator<BlockPos> getRegion(World world, BlockPos corePos, Direction dir);

    @Nullable
    StructureMatch findMatch(World world, BlockPos corePos, Direction dir);

    Collection<Pair<Vec3i, StructureBlockMatcher>> getVisualization();

}

File: src/main/kotlin/st/evening/mc/cbtweaker/structure/StructureMatcher.kt
package st.evening.mc.cbtweaker.structure

import net.minecraft.util.math.BlockPos
import net.minecraft.world.World
import st.evening.mc.cbtweaker.multiblock.MultiBlockType
import st.evening.mc.cbtweaker.util.Identifiable
import st.evening.mc.prelude.api.data.tjson.JsonPath
import st.evening.mc.prelude.api.data.tjson.TJson
import st.evening.mc.prelude.api.util.world.BlockSide

interface StructureMatcher<M> {
    fun findMatch(
        world: World,
        corePos: BlockPos,
        front: BlockSide,
        prev: M?,
        changedBlocks: Collection<BlockPos>
    ): StructureMatch<M>

    fun getVisualization(data: M?): StructureVisualization
}

interface StructureMatcherType<M> : Identifiable {
    context(_: JsonPath)
    fun loadMatcher(mbType: MultiBlockType<*>, dto: TJson.Object): StructureMatcher<M>
}
